Add Initial set of models

arthuroe opened the pull request:
References: https://github.com/quicksprout/naas-api-client-python/pull/8

### Tasks

Adds the following models
- [x] Subscriber model
- [x] Subscriber email address model
- [x] Campaign model
- [x] Account model
- [x] Account setting model
- [x] Pagination model
- [x] Account smtp setting model
- [x] Campaign email template
- [x] Email Notification Delivery model
- [x] Email Notification Status model
- [x] Email notification model
- [x] Error item model
- [x] Error model
- [x] Account settings model

arthuroe committed:
Add subscriber model

arthuroe committed:
Add subscriber email address model

arthuroe committed:
Add campaign model

arthuroe committed:
Add account model

arthuroe committed:
Add account setting model

arthuroe committed:
Add pagination model

arthuroe committed:
Add Account smtp setting model

arthuroe committed:
Add campaign email template model

arthuroe committed:
Add email notification delivery model

arthuroe committed:
Add email notification status model

arthuroe committed:
Add email notification model

arthuroe committed:
Add error item model

arthuroe committed:
Add error model

arthuroe committed:
Add account settings model

nateklaiber commented on naas/models/account_settings.py:
```diff
@@ -0,0 +1,25 @@
+import logging
+
+from requests import AccountSettings
```
@arthuroe - this gave me an error, as there is no `AccountSettings` inside of `requests`.

nateklaiber commented on naas/models/account.py:
```diff
@@ -0,0 +1,49 @@
+        if self.account_setting_attributes():
+            return AccountSetting(self.account_setting_attributes())
+        else:
+            return AccountSettings.retrieve()
```
This revealed a previous error for `AccountSettings`. This delegates to the `AccountSettings` model, which has a static method trying to make a request to `AccountSettings.retrieve()` (But there is no request). Just making note of that here.

nateklaiber commented on naas/models/campaign.py:
```diff
@@ -0,0 +1,58 @@
+from naas.models import Links, CampaignEmailTemplates
```
This gave me an error as there is no `CampaignEmailTemplates` inside of the models.

nateklaiber commented on naas/models/campaign.py:
```diff
@@ -0,0 +1,58 @@
+                self.campaign_email_templates_attributes())
+        else:
+            return CampaignEmailTemplates.list_by_project_id_and_campaign_id(
+                self.project_id(), self.id())
```
Related to error above, there is no `CampaignEmailTemplates` model.

nateklaiber commented on naas/models/email_notification.py:
```diff
@@ -0,0 +1,133 @@
+    Links, EmailNotificationStatuses, EmailNotificationDeliveries
+)
+from naas.requests import EmailNotificationStatuses as notification_status
+
```
Error here as there is no `EmailNotificationStatuses` or `EmailNotificationDeliveries` model.

nateklaiber commented on naas/models/error.py:
```diff
@@ -0,0 +1,34 @@
+from nass.models import ErrorItems
```
Should be `naas.models`.

There is no `ErrorItems` model.

nateklaiber commented on naas/models/__init__.py:
```diff
@@ -12,3 +12,17 @@
+from naas.models.error_item import ErrorItem
+from naas.models.error import Error
+from naas.models.pagination import Pagination
+from naas.models.subscriber_email_addreses import SubscriberEmailAddress
```
Should be `subscriber_email_address`

nateklaiber committed:
Fix spelling of `subscriber_email_address`

nateklaiber committed:
Fix spelling of `naas` (I do this all the time)

nateklaiber reviewed (changes requested):
Hi @arthuroe - I've found a few things while trying to work with this in the console. Some of these are spelling issues, which I went ahead and pushed the changes. There are some other areas to review:

* Models always have a _collection_ (`Projects`) and _instance_ (`Project`) model.
  * The collection object is responsible for implementing enumerable and allowing clients to work with the collection itself.
  * The instance object is responsible for instantiating the domain model for that specific entity.
* Requests are typically either singular or plural (depending on the entity type). Their only job is to provide methods to make the requests with arguments. They will always return the raw **Response**. The domain models (above) then have a static method that matches those request methods, only they return the collection or instance domain model (not the raw **Response**). 
* The `Client` itself has helper methods to be used throughout the client itself. Things like `api_host`, `access_token`, `default_headers`, `logger`, etc.

I'll try and work around some of these issues and test out the domain models themselves that you've added.

Let me know if there are any questions!

nateklaiber committed:
Return the `first_name()` and `last_name()`

This was previously returning a bound method and
not the string name as expected.

nateklaiber committed:
Model example for Subscriber

* Updated to `email_address` from `email`
* Ensured the timestamps were coerced to `datetime` objects

nateklaiber commented on naas/models/subscriber.py:
```diff
@@ -29,17 +31,19 @@ def full_name(self):
-        return self.attributes.get('updated_at')
+        return datetime.datetime.strptime(
+            self.attributes.get('updated_at'), '%Y-%m-%dT%H:%M:%S%z')
 
```
@arthuroe - Almost all of the models will include these timestamps (`created_at`, `updated_at`), so anytime we have a `Date`, `Time`, or `DateTime` object we will want to coerce it properly. This way the consumer can use date/time libraries to work with and format the dates.

This is helpful to:

* Format the dates as they choose
* At the collection level, the client can implement ranges of 'find all records between date X and Y'

nateklaiber commented on naas/models/subscriber.py:
```diff
@@ -29,17 +31,19 @@ def full_name(self):
-        return self.attributes.get('email')
+    def email_address(self):
+        """Returns the subscriber email address"""
+        return self.attributes.get('email_address')
```
I went ahead and updated this as I was testing out the model.

nateklaiber committed:
SubscriberEmailAddress timestamp updates

The `confirmed_at` is optional and might be `None`

nateklaiber commented on naas/models/subscriber_email_address.py:
```diff
@@ -35,15 +37,21 @@ def confirmation_code(self):
+          return datetime.datetime.strptime(
+            self.attributes.get('confirmed_at'), '%Y-%m-%dT%H:%M:%S%z')
+        except TypeError as e:
+          None
```
@arthuroe - This is an example of an _optional_ timestamp. An email address might not yet be _confirmed_, and so this would be `null` in the database. I tested using a record that has a `null` entry in the DB. As expected, it raised the `TypeError`, which we want to suppress here.

The other option is that we have a separate attribute of `confirmed_at_value` - which is the representation that comes from the API. Then we check to see if that is `None`. If it is not, then we coerce to the `datetime`.

Thoughts?

arthuroe replied:
@nateklaiber I think the latter could do.
```
def confirmed_at(self):
    if self.attributes.get('confirmed_at'):
        return datetime.datetime.strptime( 
                self.attributes.get('confirmed_at'), '%Y-%m-%dT%H:%M:%S%z')
```
`None` is implicitly returned where the check is not passed.

nateklaiber replied:
The main reason I use exception handling in the `ruby` gem is to handle the case where an invalid value is passed for one of the dates. I use test cases like passing in `nil`, `random` (as a strong), or `23` (as an Integer). The exception handling captures any parse error and returns `nil` (`None`, here), whereas a condition check could lead to a false positive. `self.attributes.get('confirmed_at')` might have a value that is not parse-able by `strptime`. 

Technically this should never happen, as they will always be built from the API - but when I build unit tests I try and protect from any situation. 

I'm good to follow your lead here, just wanted to provide some more context.

nateklaiber committed:
Switch to `iso8601` dependency

This prevents us from having to always use the full
strptime format.

nateklaiber commented on naas/models/subscriber_email_address.py:
```diff
@@ -35,23 +35,24 @@ def confirmation_code(self):
         """Returns the created at timestamp value"""
-        return datetime.datetime.strptime(
-            self.attributes.get('created_at'), '%Y-%m-%dT%H:%M:%S%z')
+        return iso8601.parse_date(self.attributes.get('created_at'))
```
@arthuroe - any thoughts here on using the [iso8601](https://pypi.org/project/iso8601/) dependency? This seems to be a cleaner abstraction for our use cases. These timestamps will always be returned in that format from the API.

arthuroe replied:
Does this change affect what you mentioned below?

> @arthuroe - Almost all of the models will include these timestamps (created_at, updated_at), so anytime we have a Date, Time, or DateTime object we will want to coerce it properly. This way the consumer can use date/time libraries to work with and format the dates.
> 
> This is helpful to:
> 
> Format the dates as they choose
> At the collection level, the client can implement ranges of 'find all records between date X and Y'

nateklaiber replied:
@arthuroe - The topic is still the same. This dependency wraps up the initial `datetime` calls and eliminates us having to call `strptime` with the string format. In terms of exception handling, the conversation is still the same.

arthuroe replied:
Alright. Am okay going forward with `iso8601` dependency.

arthuroe commented:
@nateklaiber I had started with the instance models, moving towards the collections. The requests will be added by `Allan` though I can add place holders as I reference each of them.

nateklaiber commented:
Sounds good @arthuroe - will keep an eye on this PR and review when I can test with the data in collections and instances. Does that work?

arthuroe commented:
Yes, that works @nateklaiber. Was off on vacation last week, will let you know when the remaining collections are added.

arthuroe committed:
Add Account settings model

arthuroe committed:
Add accounts and campaign email templates models

nateklaiber commented:
Sounds good @arthuroe! Hope you had a great vacation, and welcome back!

arthuroe committed:
Add campaigns and email notification deliveries models

arthuroe committed:
Add email notifications and notification statuses models

arthuroe committed:
Add subscribers model

arthuroe committed:
Add email notification basic and subscriber email addresses

arthuroe committed:
Add error items model

arthuroe committed:
Add doc strings

arthuroe commented:
@nateklaiber All collections are now up. You can review when you get the time.

arthuroe committed:
Add missing import

arthuroe committed:
unify import format

arthuroe committed:
change request imports

arthuroe committed:
Remove requests import

arthuroe committed:
Fix attribute error in links

arthuroe committed:
Fix logging config

nateklaiber commented:
@arthuroe @muhallan - Here is an example of the request and response for the **Email Notification Basic**. 

> This is from my local server

```
curl -X "POST" "http://api.dev.naas.com/email-notifications/basic" \
     -H 'Accept: application/vnd.naas.json; version=1' \
     -H 'Accept-Encoding: gzip' \
     -H 'Authorization: Bearer 9c4bf569de7a48154c9399ac820643d6051b24509b91366b5876bed0d987' \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{
  "email_notification_basic": {
    "campaign_email_template_id": "my-first-campaign-template",
    "email_address": "nate+r2@deviceindependent.com",
    "content": {
      "name": "Nate Klaiber"
    },
    "campaign_id": "my-first-campaign",
    "tags": {
      "awesome": "dsf",
      "batch_id": "12"
    },
    "project_id": "my-first-project"
  }
}'
```

And here is the response:

```json
{
  "links": [
    {
      "title": "Email Notifications",
      "href": "http://naas-api-local.deviceindependent.com/email-notifications",
      "rel": "self",
      "templated": false
    }
  ],
  "data": {
    "id": 612,
    "account_smtp_setting_id": "my_testing",
    "campaign_email_template_id": "my-first-campaign-template",
    "subscriber_email_address_id": 520,
    "content": {
      "name": "Nate Klaiber"
    },
    "subject": "TESTING",
    "from_email_address": "nate@deviceindependent.com",
    "from_name": "Nate Klaiber",
    "to_email_address": "nate+r2@deviceindependent.com",
    "to_name": " ",
    "html_body": "Welcome, Nate Klaiber!",
    "text_body": "Welcome, Nate Klaiber!",
    "sent_count": 0,
    "last_sent_at": null,
    "is_sent": false,
    "is_processing": false,
    "is_deliverable": true,
    "checksum": "d5e1c5b6ab7f0d0f8690d4e3a3ecc6c6",
    "created_at": "2019-05-06T17:50:05Z",
    "updated_at": "2019-05-06T17:50:05Z",
    "links": [
      {
        "title": "Detail",
        "href": "http://naas-api-local.deviceindependent.com/email-notifications/612",
        "rel": "self",
        "templated": false
      },
      {
        "title": "Preview",
        "href": "http://naas-api-local.deviceindependent.com/email-notifications/612/preview",
        "rel": "http://naas-api-local.deviceindependent.com/rels/email-notification-preview",
        "templated": false
      },
      {
        "title": "Deliveries",
        "href": "http://naas-api-local.deviceindependent.com/email-notifications/612/deliveries",
        "rel": "http://naas-api-local.deviceindependent.com/rels/email-notification-deliveries",
        "templated": false
      },
      {
        "title": "Deliver",
        "href": "http://naas-api-local.deviceindependent.com/email-notifications/612/deliver",
        "rel": "http://naas-api-local.deviceindependent.com/rels/email-notification-deliver",
        "templated": false
      }
    ]
  }
}
```
